src/main/java/org/mariadb/jdbc/internal/failover/impl/AuroraListener.java

```diff
@@ -154,47 +206,245 @@ public void reconnectFailedConnection(SearchFilter searchFilter) throws QueryExc
+                if (!currentHosts.contains(endpoint)) {
+                    HostAddress newHostAddress = new HostAddress(endpoint, port, null);
+                    addresses.add(newHostAddress);
+                }
+            }
+        }
+    }
+
+    /**
+     * Looks for the current master/writer instance via the secondary protocol if it is found within 3 attempts.
+     * Should it not be able to connect, the host is blacklisted and null is returned.
+     * Otherwise, it will open a new connection to the cluster endpoint and retrieve the data from there.
      *
      * @param secondaryProtocol the current secondary protocol
      * @param loopAddress       list of possible hosts
      * @return the probable master address or null if not found
      */
     public HostAddress searchByStartName(Protocol secondaryProtocol, List<HostAddress> loopAddress) {
         if (!isSecondaryHostFail()) {
-            MariaSelectResultSet queryResult = null;
-            try {
-                proxy.lock.lock();
+            int checkWriterAttempts = 3;
+            HostAddress currentWriter = null;
+
+            do {
                 try {
-                    SingleExecutionResult executionResult = new SingleExecutionResult(null, 0, true, false);
-                    secondaryProtocol.executeQuery(executionResult,
-                            "select server_id from information_schema.replica_host_status where session_id = 'MASTER_SESSION_ID'",
-                            ResultSet.TYPE_FORWARD_ONLY);
-                    queryResult = executionResult.getResult();
-                    queryResult.next();
-                } finally {
-                    proxy.lock.unlock();
+                    currentWriter = searchForMasterHostAddress(secondaryProtocol, loopAddress);
+                } catch (QueryException qe) {
+                    if (proxy.hasToHandleFailover(qe) && setSecondaryHostFail()) {
+                        addToBlacklist(secondaryProtocol.getHostAddress());
+                        return null;
+                    }
                 }
-                String masterHostName = queryResult.getString(1);
-                for (int i = 0; i < loopAddress.size(); i++) {
-                    if (loopAddress.get(i).host.startsWith(masterHostName)) {
-                        return loopAddress.get(i);
+                checkWriterAttempts--;
+            } while (currentWriter == null && checkWriterAttempts > 0);
+
+            // Handling special case where no writer is found from secondaryProtocol
+            if (currentWriter == null && getClusterHostAddress() != null) {
+                AuroraProtocol tempProtocol = AuroraProtocol.getNewProtocol(getProxy(), getUrlParser());
+                tempProtocol.setHostAddress(getClusterHostAddress());
+                try {
+                    tempProtocol.connect();
+                    return searchForMasterHostAddress(tempProtocol, loopAddress);
```

Comment:
same remark than previous one : goal is to found current master host to directly connect to the current master instance. if master is connected, no need to try a connection after ward, just set current master to this protocol